### ISSUE TITLE
Visual Bug: all components get crammed to the left

### ISSUE THREAD

**Issue by subwaystation** [graph-genome/Schematize#54]
(edited by Josiah Seaman)
Christine first found this bug in https://github.com/graph-genome/Schematize/pull/49#issuecomment-609968036  It [shows up in both branches](https://github.com/graph-genome/Schematize/pull/46#issuecomment-609466064) from Bonface and Jacob, so it seems they're exposing an underlying issue.

Cause appears to be: all component x gets set to 0.  Re-renders fix the issue, so this is likely a render before x is set.

#### Ways to reproduce this bug:
* Changing the row height to 6
* Use Path Position Jump button: set Path 262 and index 120 then click "Jump"
* In 1db3e3a it's possible to mouse over a LinkColumn to trigger this, that doesn't trigger it in 
3bbb8b5

The commonality is all of these events cause a re-render without necessarily updating the data.

![image](https://user-images.githubusercontent.com/10184961/78594626-f153c980-7848-11ea-84e6-6eeb96d7e87f.png)
*What we're seeing is all the components and their matrices piled up on the left.*

![image](https://user-images.githubusercontent.com/2256148/78602313-7d4afe80-7813-11ea-944b-2b3a8f3d9421.png)
*It's possible to get individual components to re-render by clicking on them on the left. Triggering a full re-render (move arrows) fixes the visual glitch. For example change "Pangenome Bin Position:" to 2.*

**Comment by ChristineSeaman** [comment]
Notes from the other issue in another branch: 

1. Issue: Changed the path position to "262" : 120 and click jump re-rendered. Vast majority of information disappeared from the matrix, but seemed to stack up on top of themselves on the left hand of the screen. If the user clicks on the columns that are displaying then they start populating the matrix one column at a time. Screen shots below.
2. When testing that the cells in the matrix match the color of their corresponding arrow, at least one link in the green section did not have any related cells displayed. Activating the Vertical Compression then displays a single middle green cells/row at the top. It is possible that some of the information is not displaying at the bottom of the screen scroll in the default views.
3. When mouse-overing the cells (grey blocks) it displays the cell coordinates, but not the corresponding labels (e.g. Coverage, Inversion).

**Comment by josiahseaman** [comment]
For anyone trying to solve this: I believe this is a symptom of an earlier comment: "Also the mobx updates invoked by updateMaxHeight() and resetRenderStats() are not ideal since the information needed at the beginning of the render isn't available until the end of the render frame.
As of 8a9d925 it takes two frames to get the desired behavior. Please could you apply your newfound mobx knowledge to updating state in the right order? I'm hoping it's possible in a single pass. If not, then most likely the vertical compression heights will need to be precalculated before the render. That will require a significant refactor, so I'm hoping it's not necessary. This is also a useful use case to test how much render metadata we can put in the state tree." #7 

((Are you allowed to use a colon twice in a sentence?))

**Comment by subwaystation** [comment]
Currently going through a bunch of previous commits to find the last one, where this issue did not appear.

**Comment by subwaystation** [comment]
The first time I am seeing the issue is https://github.com/graph-genome/Schematize/commit/9a15962739f2086bc591dfe8241aed2446d72653. Which is weird. Only the READM.md is updated. I am really confused.

**Comment by subwaystation** [comment]
One commit earlier https://github.com/graph-genome/Schematize/commit/9dd54119a57d9457f420909de03631301b03de4a works.

**Comment by josiahseaman** [comment]
One thing I've noticed in testing is sometime you need to open developer tools -> Network -> disable caching.  The browser is really stubborn about switching slowly, especially when you haven't restarted npm, so you can get weird results that way. There's an underlying vulnerability in our state machine, and lots of little issues either expose or hide the same underlying problem.

**Comment by subwaystation** [comment]
I am always doing a `npm i && npm start` after each checkout. But let me also disable the network caching.

**Comment by subwaystation** [comment]
Thanks for the tip, but it did not change anything.

**Comment by subwaystation** [comment]
So my plan would be to checkout the last functional commit and add code from `master` into it, as long as it works. Then we will have a little bit more stable master.

**Comment by josiahseaman** [comment]
I've asked Christine to collect evidence for when and in what branches this crops up, then we can make the code changes in master.  We definitely shouldn't revert the last few weeks of progress to fix a bug.  The underlying issue is an inconsistency in order of events. It should always look like this:

1. BeginEndBin changed (or startup), fetch URLs from bin2file.json
2. Data loaded into schematize PangenomeSchematic.components models
3. recalcXLayout assigns x values for each column: LinkColumn, padding, xstart in each component
4. X values set in components
5. X values set in Links
6. calculateLinkElevations() does vertical stacking and updates [topOffset](https://github.com/graph-genome/Schematize/blob/1db3e3a2ed9b783172f466bdc81221c0fd4ee9e2/src/App.js#L111)
7. Only after all the layout is done, render() happens with the correct topOffset and correct X values for components.  Components and Links line up because they're both on the same version of x values from recalcXLayout()

Somewhere we've got cases where an update is jumping in the middle.  More likely, the mobx-state-tree updates are asynchronous so the render() loop queues some changes that affect the next render() not the current one.

**Comment by ChristineSeaman** [comment]
Upon testing, it appears that the initial rendering is correct. Then every rendering after that the X values get set and then it renders, then Y values get set but it does not trigger a rendering. It is possible that it is performing steps 4, then 5, then 7, then maybe step 6. Or maybe not.

**Comment by BonfaceKilz** [comment]
I've figured out the source of the bug.. You'll notice that page renders twice, and on the second render, things get skewed to the left. This was not logged before so this is how to reproduce it:
- Run `npm start`
- Wait for anywhere between 10s to 30s
I run:
-  Intel Core i5 (2nd Gen) 2557M / 1.7 GHz
- 8GB RAM
so time may vary for you.

To figure out what was causing the render[0], I updated `componentDidUpdate` in the App.js with this:
```
    Object.entries(this.props).forEach(
      ([key, val]) =>
        prevProps[key] !== val && console.log(`Prop '${key}' changed`)
    );
    if (this.state) {
      Object.entries(this.state).forEach(
        ([key, val]) =>
          prevState[key] !== val && console.log(`State '${key}' changed`)
      );
    }
```
which basically logs any updated state to the console. Our culprit is `this.setState({ highlightedLink: linkRect });`. I'll submit a working patch in a few.

[0] https://stackoverflow.com/questions/41004631/trace-why-a-react-component-is-re-rendering

**Comment by josiahseaman** [comment]
![image](https://user-images.githubusercontent.com/2256148/79006921-59d7ca80-7b17-11ea-9151-93fc237ae3ec.png)
Unfortunately I wasn't able to find a reproducible path to show this bug, but there are moments where it's on screen before it is refreshed.

The current PR isn't perfect, but it's a tangible improvement on master branch.  I'm leaving a note here to remind us this fix may cause additional re-renders that could slow the browser.  We'll postpone that until we get to a browser optimization step.  I believe this issue should remain open for now.
